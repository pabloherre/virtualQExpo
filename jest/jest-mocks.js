import mockAsyncStorage from '@react-native-community/async-storage/jest/async-storage-mock';

const mockfeathersMethods = {
  create: jest.fn(),
  update: jest.fn(),
  find: jest.fn(),
  get: jest.fn(),
  patch: jest.fn(),
  remove: jest.fn()
};
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.requireActual('./mocks');
jest.mock('@react-native-community/async-storage', () => mockAsyncStorage);
jest.mock('expo-barcode-scanner');
jest.mock('expo-location');
jest.mock('react-native-maps');
jest.mock('../src/setup/feathersClient', () => {
  return {
    __esModule: true,
    default: jest.fn(),
    userApi: mockfeathersMethods,
    appointmentApi: mockfeathersMethods,
    businessApi: mockfeathersMethods,
    categoryApi: mockfeathersMethods,
    turnApi: mockfeathersMethods,
    authenticate: jest.fn(),
    reAuthenticate: jest.fn()
  };
});
jest.mock('../src/services/user/User.service');
jest.mock('../src/modules/auth/Auth.service');
jest.mock('../src/common/sideMenu/SideMenu.service');
jest.genMockFromModule('../src/setup/store');
