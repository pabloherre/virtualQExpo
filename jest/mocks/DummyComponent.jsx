import React, { Component } from 'react';
import { View } from 'react-native';

export default class DummyComponent extends Component {
  render() {
    return <View></View>;
  }
}
