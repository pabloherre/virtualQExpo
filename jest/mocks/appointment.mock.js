export const appointment = {
  id: 1,
  date: 'Today',
  business: 'Pharmacy',
  appointmentNumber: '234234',
  appointmentCountdown: 23,
  latitude: -32.8913568,
  longitude: -68.8567443,
  address: 'Av. Cabildo 2035',
  city: 'Belgrano, CABA',
  email: 'somename@randomemail.com'
};
