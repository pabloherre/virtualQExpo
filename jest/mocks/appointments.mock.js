export const appointments = [
  {
    id: 1,
    date: 'Today',
    business: 'Pharmacy',
    appointmentNumber: '234234',
    appointmentCountdown: 23,
    latitude: -32.8913568,
    longitude: -68.8567443
  },
  {
    id: 2,
    date: 'Today',
    business: 'Supermarket',
    appointmentNumber: '2B4354',
    appointmentCountdown: 322,
    latitude: -32.8913568,
    longitude: -68.8567443
  },
  {
    id: 3,
    date: 'Tomorrow',
    business: 'Cinema',
    appointmentNumber: '234354',
    appointmentCountdown: 322,
    latitude: -32.8913568,
    longitude: -68.8567443
  },
  {
    id: 4,
    date: 'Tomorrow',
    business: 'Pool',
    appointmentNumber: '234354',
    appointmentCountdown: 322,
    latitude: -32.8913568,
    longitude: -68.8567443
  }
];
