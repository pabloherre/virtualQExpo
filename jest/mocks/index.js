export * from './appointment.mock';
export * from './appointments.mock';

