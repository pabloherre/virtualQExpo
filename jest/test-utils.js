import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockStore = configureStore([]);
function connectedRender(ui, initialState) {
  const store = mockStore(initialState);
  component = renderer.create(<Provider store={store}>{ui}</Provider>);
  return component;
}

// override connectedRender method
export { connectedRender };
