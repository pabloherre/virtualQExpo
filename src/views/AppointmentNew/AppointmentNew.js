import React, { Component } from 'react';
import { connect } from 'react-redux';
import Typography from '../../common/typography/Typography';
import { safeArea } from '../../styles/common.styles';
import { SafeAreaView, TextInput, TouchableOpacity, View, Text } from 'react-native';
import { BarCodeScanner } from 'expo-barcode-scanner';
import { RoundedButton } from '../../common/buttons/';
import { AntDesign } from '@expo/vector-icons';
import MapView, { Marker } from 'react-native-maps';
import * as Location from 'expo-location';

class AppointmentNew extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentLocation: null,
      searchValue: '',
      hasPermissions: false
    };
  }

  async componentDidMount() {
    let { status } = await Location.requestPermissionsAsync();
    if (status === 'granted') {
      let location = await Location.getCurrentPositionAsync({ accuracy: Location.Accuracy.High });
      console.log(location);
      this.setState({ currentLocation: location });
    }
  }

  handleCode = () => {
    this.props.navigation.navigate('AppointmentDetails', { appointment: this.props.appointments[0] }); // TODO: reserve appointment for this business
  };

  render() {
    const { hasPermissions, searchValue, currentLocation } = this.state;
    return (
      <SafeAreaView style={safeArea}>
        <Typography>Type the store code or scan the QR code</Typography>

        <View style={{ flexDirection: 'row' }}>
          <View
            style={{
              flexDirection: 'row',
              flex: 2,
              alignItems: 'center',
              borderColor: 'gray',
              justifyContent: 'space-between',
              borderWidth: 1,
              borderRadius: 30,
              marginRight: 5,
              paddingHorizontal: 10
            }}
          >
            <TextInput style={{ flex: 8 }} value={searchValue} onChangeText={text => this.setState({ searchValue: text })} label="Code" />
            <TouchableOpacity style={{ flex: 1 }} onPress={this.handleCode}>
              <AntDesign name="right" size={20} color="#f77027" />
            </TouchableOpacity>
          </View>
          <RoundedButton
            buttonStyle={{ flex: 1 }}
            label={'SCAN'}
            onPress={async () => {
              if (!hasPermissions) {
                const { status } = await BarCodeScanner.requestPermissionsAsync();
                if (status === 'granted') {
                  this.props.navigation.navigate('QRCode');
                }
              }
            }}
          />
        </View>

        <View
          style={{
            width: '100%',
            borderBottomWidth: 1,
            borderBottomColor: 'lightgray',
            marginVertical: 20,
            borderStyle: 'dashed'
          }}
        />

        <Typography>You can also find stores near you</Typography>
        {currentLocation && (
          <View style={{ borderRadius: 30, overflow: 'hidden', flex: 1 }}>
            <MapView
              style={{ padding: 20, flex: 1 }}
              initialRegion={{
                latitude: currentLocation.coords.latitude,
                longitude: currentLocation.coords.longitude,
                latitudeDelta: 0.0922,
                longitudeDelta: 0.0421
              }}
            >
              {this.props.appointments.map(appointment => {
                return (
                  <Marker
                    onPress={this.handleCode}
                    coordinate={{ latitude: appointment.latitude, longitude: appointment.longitude }}
                    title={appointment.business}
                    description={appointment.business}
                  />
                );
              })}
            </MapView>
          </View>
        )}
      </SafeAreaView>
    );
  }
}

function mapStateToProps(state) {
  return {
    appointments: state.appointment.appointments
  };
}

export default connect(mapStateToProps)(AppointmentNew);
