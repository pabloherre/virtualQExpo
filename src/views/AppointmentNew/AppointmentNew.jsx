import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Typography from '../../common/typography/Typography';
import { safeArea } from '../../styles/common.styles';
import { TextInput, TouchableOpacity, View } from 'react-native';
import { BarCodeScanner } from 'expo-barcode-scanner';
import { RoundedButton } from '../../common/buttons';
import { AntDesign } from '@expo/vector-icons';
import MapView, { Marker } from 'react-native-maps';
import * as Location from 'expo-location';
import { colors } from '../../../theme';
import debounce from '@react-navigation/stack/src/utils/debounce';
import TurnService from '../../services/turn/Turn.service';
import { showMessage } from 'react-native-flash-message';

export class AppointmentNew extends Component {
  constructor(props) {
    super(props);
    this.state = {
      latitude: 0,
      longitude: 0,
      searchValue: '',
      hasPermissions: false,
      business: []
    };
  }

  async componentDidMount() {
    let { status } = await Location.requestPermissionsAsync();
    if (status === 'granted') {
      let location = await Location.getCurrentPositionAsync({ accuracy: Location.Accuracy.High });
      this.setState({ latitude: location.coords.latitude, longitude: location.coords.longitude }, () =>
        this.findTurnsByRegion({ latitude: location.coords.latitude, longitude: location.coords.longitude, latitudeDelta: 0.0922 })
      );
    }
  }

  getBusinessByTurns = () => {
    return this.props.turns.reduce((acc, turn) => {
      return acc.concat(turn.business);
    }, []);
  };

  handleCode = async code => {
    const turn = await TurnService.findBusinessByCode(code);
    if (!turn) {
      showMessage({
        message: 'Invalid code',
        description: 'Business not found',
        type: 'danger',
        icon: 'danger'
      });
    } else {
      const appointment = {
        turn,
        date: Date.now().toString()
      };
      this.props.navigation.navigate('AppointmentDetails', { appointment });
    }
  };

  findTurnsByRegion = async region => {
    const { latitude, longitude } = this.state;

    TurnService.findNearTurns({
      latitude,
      longitude,
      delta: region.latitudeDelta
    });
  };

  handleChangeRegion = newRegion => {
    this.debouncedHandleChangeRegion(newRegion);
  };

  debouncedHandleChangeRegion = debounce(newRegion => {
    this.setState({ latitude: newRegion.latitude, longitude: newRegion.longitude }, () => this.findTurnsByRegion(newRegion));
  }, 1000);

  render() {
    const { hasPermissions, searchValue, latitude, longitude } = this.state;
    return (
      <View style={safeArea}>
        <Typography>Type the store code or scan the QR code</Typography>

        <View style={{ flexDirection: 'row', marginTop: 20 }}>
          <View
            style={{
              flexDirection: 'row',
              flex: 2,
              alignItems: 'center',
              borderColor: colors.border,
              justifyContent: 'space-between',
              borderWidth: 1,
              borderRadius: 30,
              marginRight: 5,
              paddingHorizontal: 10
            }}
          >
            <TextInput style={{ flex: 8 }} value={searchValue} onChangeText={text => this.setState({ searchValue: text })} label="Code" />
            <TouchableOpacity id="searchCode" style={{ flex: 1 }} onPress={() => this.handleCode(searchValue)}>
              <AntDesign name="right" size={20} color={colors.primary} />
            </TouchableOpacity>
          </View>

          <RoundedButton
            buttonStyle={{ flex: 1 }}
            label={'SCAN'}
            onPress={async () => {
              if (!hasPermissions) {
                const { status } = await BarCodeScanner.requestPermissionsAsync();
                if (status === 'granted') {
                  this.props.navigation.navigate('QRCode');
                }
              }
            }}
          />
        </View>

        <View
          style={{
            width: '100%',
            borderBottomWidth: 1,
            borderBottomColor: 'lightgray',
            marginVertical: 20,
            borderStyle: 'dashed'
          }}
        />

        <Typography>You can also find stores near you</Typography>

        <View
          style={{ borderRadius: 30, overflow: 'hidden', flex: 1, marginTop: 20, borderColor: colors.border, borderWidth: Boolean(latitude) ? 0 : 1 }}
        >
          {Boolean(latitude) && Boolean(longitude) && (
            <MapView
              style={{ padding: 20, flex: 1 }}
              initialRegion={{
                latitude,
                longitude,
                latitudeDelta: 0.0922,
                longitudeDelta: 0.0421
              }}
              onRegionChange={reg => this.handleChangeRegion(reg)}
            >
              {this.getBusinessByTurns().map(business => {
                return (
                  <Marker
                    key={business._id}
                    onCalloutPress={() => {
                      this.handleCode(business.code);
                    }}
                    coordinate={{
                      latitude: business.location.coordinates[1],
                      longitude: business.location.coordinates[0]
                    }}
                    title={business.name}
                    description={business.address}
                  />
                );
              })}
            </MapView>
          )}
        </View>
      </View>
    );
  }
}

AppointmentNew.propTypes = {
  colors: PropTypes.shape({
    primary: PropTypes.string
  })
};

AppointmentNew.defaultProps = {
  colors: colors
};

function mapStateToProps(state) {
  return {
    turns: state.turn.data
  };
}

export default connect(mapStateToProps)(AppointmentNew);
