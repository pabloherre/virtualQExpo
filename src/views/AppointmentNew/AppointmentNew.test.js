import React from 'react';
import { AppointmentNew } from './AppointmentNew';
import { shallow } from 'enzyme';
import { connectedMount } from '../../../jest/test-utils';
import { BarCodeScanner } from 'expo-barcode-scanner';
import * as Location from 'expo-location';
import MapView, { Marker } from 'react-native-maps';

import { appointments } from '../../../jest/mocks';

let wrapper;
const mockProps = {
  appointments,
  navigation: {
    navigate: jest.fn()
  }
};

describe('<AppointmentNew>', () => {
  beforeEach(() => {
    wrapper = shallow(<AppointmentNew {...mockProps} />);
  });

  afterEach(() => {
    Location.requestPermissionsAsync.mockClear();
    Location.getCurrentPositionAsync.mockClear();
    mockProps.navigation.navigate.mockClear();
  });
  it('should render correctly', async () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should ask for location permission to the user and dont try to get location if permission not granted', async () => {
    Location.requestPermissionsAsync.mockReturnValueOnce({});
    //mount component again so we mock the response of Location reques
    //on componentDidMountfor this test
    wrapper = await shallow(<AppointmentNew {...mockProps} />);
    expect(Location.requestPermissionsAsync).toHaveBeenCalled();
    expect(Location.getCurrentPositionAsync).toBeCalledTimes(0);
  });

  it('should ask for location permission to the user and get location if granted', async () => {
    Location.requestPermissionsAsync.mockReturnValueOnce({ status: 'granted' });
    //mount component again so we mock the response of Location reques
    //on componentDidMountfor this test
    wrapper = await shallow(<AppointmentNew {...mockProps} />);
    expect(Location.requestPermissionsAsync).toHaveBeenCalled();
    expect(Location.getCurrentPositionAsync).toHaveBeenCalled();
  });
  it('should have an input for the appointment code', async () => {
    expect(wrapper.find('Component[label="Code"]')).toExist();
  });

  it('should have an scan button dont redurects you if permission is not granted', async () => {
    BarCodeScanner.requestPermissionsAsync.mockReturnValueOnce({});
    await wrapper.find('RoundedButton').simulate('press');
    expect(BarCodeScanner.requestPermissionsAsync).toHaveBeenCalled();
    expect(wrapper.instance().props.navigation.navigate).toBeCalledTimes(0);
  });

  it('should have an scan button that redirects you to the camera view  if permission granted', async () => {
    BarCodeScanner.requestPermissionsAsync.mockReturnValueOnce({ status: 'granted' });
    await wrapper.find('RoundedButton').simulate('press');
    expect(BarCodeScanner.requestPermissionsAsync).toHaveBeenCalled();
    expect(wrapper.instance().props.navigation.navigate).toHaveBeenCalled();
  });

  it('should update the searchValue state when writting on Code input', async () => {
    wrapper.find('Component[label="Code"]').props().onChangeText('code');
    expect(wrapper.state('searchValue')).toBe('code');
  });

  it('should redirect you to the appointment after processing the appointment', async () => {
    wrapper.find('#searchCode').simulate('press');
    expect(wrapper.instance().props.navigation.navigate).toHaveBeenCalled();
  });
});
