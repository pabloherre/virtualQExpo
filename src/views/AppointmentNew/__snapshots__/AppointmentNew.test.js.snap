// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppointmentNew> should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingHorizontal": 40,
      "paddingVertical": 20,
    }
  }
>
  <Component>
    Type the store code or scan the QR code
  </Component>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "borderColor": "rgb(216, 216, 216)",
          "borderRadius": 30,
          "borderWidth": 1,
          "flex": 2,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginRight": 5,
          "paddingHorizontal": 10,
        }
      }
    >
      <Component
        accessibilityLabel="testInputCode"
        allowFontScaling={true}
        label="Code"
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "flex": 8,
          }
        }
        testID="testInputCode"
        underlineColorAndroid="transparent"
        value=""
      />
      <ForwardRef
        accessibilityLabel="testTouchableSearchCode"
        id="searchCode"
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
        testID="testTouchableSearchCode"
      >
        <Icon
          allowFontScaling={false}
          color="#f77027"
          name="right"
          size={20}
        />
      </ForwardRef>
    </View>
    <RoundedButton
      accessibilityLabel="testTouchableScan"
      buttonStyle={
        Object {
          "flex": 1,
        }
      }
      label="SCAN"
      onPress={[Function]}
      testID="testTouchableScan"
      textStyle={Object {}}
    />
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "lightgray",
        "borderBottomWidth": 1,
        "borderStyle": "dashed",
        "marginVertical": 20,
        "width": "100%",
      }
    }
  />
  <Component>
    You can also find stores near you
  </Component>
  <View
    style={
      Object {
        "borderColor": "rgb(216, 216, 216)",
        "borderRadius": 30,
        "borderWidth": 1,
        "flex": 1,
        "marginTop": 20,
        "overflow": "hidden",
      }
    }
  />
</View>
`;
